### PR TITLE
Added ASN1 support macros needed for X509 to CHIP certificate conversion

### PR DESCRIPTION
**Pull request by emargolis**
Added ASN1 support macros needed for X509 to CHIP certificate conversion.

 #### Problem
To be able to process and operate with CHIP certificates basic ASN.1 encoding/decoding functions are needed.

 #### Summary of Changes
      * Implements ASN1Writer and ASN1Reader parsers
      * Implements mactors to operate with ASN1Writer/ASN1Reader classes
      * Implements gen_asn1oid.py python script for automatic generation of OID values
      * Added ZigBee specific attributes for CHIP Certificates:
        - ChipNodeId
        - ChipCAId
        - ChipSoftwarePublisherId

### CODE REVIEW

**Comment by mspang** [comment]
Should it be in src/crypto ?

File: src/lib/support/BUILD.gn [hunk 1]
```diff
@@ -21,6 +21,9 @@ import("${chip_root}/build/chip/chip_version.gni")
 import("${chip_root}/build/chip/tests.gni")
 import("${chip_root}/src/lib/core/core.gni")
 
+write_file("${root_gen_dir}/include/support/ASN1OID.h",
```
Comment (mspang) [inline src/lib/support/BUILD.gn:24]: This should run at build time via an `action` rule, not graph generation time via `exec_script` and `write_file`.

Generally the only things that need to call `exec_script()` for are those that affect the graph structure or command lines.

We have several examples of code generation already at build time: `buildconfig_header()`, `gdbus_library()`, `gen_flashing_script()`. This case is even simpler than those.

**Review by mspang — CHANGES_REQUESTED** [review]

**Comment by andy31415** [comment]
Could you either remove the 'fill me in' tags from the PR description or provide more details?

File: src/lib/support/ASN1.h [hunk 2]
```diff
@@ -0,0 +1,217 @@
...
+ */
+
+/**
+ *    @file
+ *      This file defines types and objects for reading and writing
+ *      Abstract Syntax Notation One (ASN.1) encoded data.
+ *
+ */
+
+#ifndef ASN1_H_
```
Comment (andy31415) [inline src/lib/support/ASN1.h:27]: `#pragma once` is generally used within CHIP now.

File: src/lib/support/ASN1.h [hunk 3]
```diff
@@ -0,0 +1,217 @@
...
+    kASN1UniversalTag_IA5String       = 22,
+    kASN1UniversalTag_UTCTime         = 23,
+    kASN1UniversalTag_GeneralizedTime = 24,
+    kASN1UniversalTag_GraphicString   = 25,
+    kASN1UniversalTag_VisibleString   = 26,
+    kASN1UniversalTag_GeneralString   = 27,
+    kASN1UniversalTag_UniversalString = 28
+};
+
+class ASN1UniversalTime
```
Comment (andy31415) [inline src/lib/support/ASN1.h:89]: A class with all public data members seem to be better representable as a struct.

File: src/lib/support/ASN1.h [hunk 4]
```diff
@@ -0,0 +1,217 @@
...
+    uint8_t Minute;
+    uint8_t Second;
+};
+
+class DLL_EXPORT ASN1Reader
+{
+public:
+    void Init(const uint8_t * buf, uint32_t len);
+
+    uint8_t Class;
```
Comment (andy31415) [inline src/lib/support/ASN1.h:105]: Could we make the data members non-public?

File: src/lib/support/ASN1.h [hunk 5]
```diff
@@ -0,0 +1,217 @@
...
+        const uint8_t * ElemStart;
+        uint32_t HeadLen;
+        uint32_t ValueLen;
+        bool IsIndefiniteLen;
+        const uint8_t * ContainerEnd;
+    };
+
+    enum
+    {
+        kMaxContextDepth = 32
```
Comment (andy31415) [inline src/lib/support/ASN1.h:138]: would `static constexpr size_t` work better than enum?

**Comment by rwalker-apple** [comment]
> Added ASN1 support macros needed for X509 to CHIP certificate conversion.
> 
> #### Problem
> <<<<<FILL ME IN - the issue this PR is intended to address>>>>>
> 
> #### Summary of Changes
> <<<<<FILL ME IN - what's in this PR>>>>>
> 
> Fixes #<<<<<FILL ME IN - issue number(s). If no issue, please create one>>>>>

Would you mind filling in the template for the PR description?  The text that starts with <<< and ends with >>> should be replaced by the contributor.  The PR description is an important input to the review process.

File: src/lib/support/BUILD.gn (hunk 1 above)
Comment (emargolis) [reply src/lib/support/BUILD.gn]: Thanks for pointing that out. I've update the script.

File: src/lib/support/ASN1.h (hunk 2 above)
Comment (emargolis) [reply src/lib/support/ASN1.h]: done

File: src/lib/support/ASN1.h (hunk 3 above)
Comment (emargolis) [reply src/lib/support/ASN1.h]: updated

File: src/lib/support/ASN1.h (hunk 4 above)
Comment (emargolis) [reply src/lib/support/ASN1.h]: done

File: src/lib/support/ASN1.h [hunk 6]
```diff
@@ -0,0 +1,217 @@
...
+    uint8_t * mWritePoint;
+    uint8_t ** mDeferredLengthList;
+
+    ASN1_ERROR EncodeHead(uint8_t cls, uint32_t tag, bool isConstructed, int32_t len);
+    ASN1_ERROR WriteDeferredLength(void);
+    static uint8_t GetLengthOfLength(int32_t len);
+    static void EncodeLength(uint8_t * buf, uint8_t lenOfLen, int32_t lenToEncode);
+};
+
+extern OID ParseObjectID(const uint8_t * encodedOID, uint16_t encodedOIDLen);
```
[the remark this answers is not in the document]
Comment (emargolis) [reply src/lib/support/ASN1.h]: done

File: src/lib/support/ASN1Config.h [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+ */
+
+/**
+ *    @file
+ *      This file defines default compile-time configuration constants
+ *      for the CHIP ASN1 subsystem.
+ *
+ */
+
+#ifndef ASN1_CONFIG_H_
```
[the remark this answers is not in the document]
Comment (emargolis) [reply src/lib/support/ASN1Config.h]: done

File: src/lib/support/ASN1Error.h [hunk 8]
```diff
@@ -0,0 +1,221 @@
...
+// codes to strings in ASN1Error.cpp, and add them to unittest
+// in test-apps/TestErrorStr.cpp
+
+/**
+ *  @}
+ */
+
+// clang-format on
+
+extern bool FormatASN1Error(char * buf, uint16_t bufSize, int32_t err);
```
[the remark this answers is not in the document]
Comment (emargolis) [reply src/lib/support/ASN1Error.h]: done

File: src/lib/support/ASN1Macros.h [hunk 9]
```diff
@@ -0,0 +1,310 @@
...
+ */
+
+#ifndef ASN1MACROS_H_
+#define ASN1MACROS_H_
+
+#include <support/CodeUtils.h>
+
+// Local variable names used by utility macros.
+//
+// Re-define these to use different names.
```
[the remark this answers is not in the document]
Comment (emargolis) [reply src/lib/support/ASN1Macros.h]: I would prefer to keep macros for now and create GH issue to review that later.

File: src/lib/support/ASN1Reader.cpp [hunk 10]
```diff
@@ -0,0 +1,481 @@
...
+ASN1_ERROR ASN1Reader::ExitContainer()
+{
+    if (mNumSavedContexts == 0)
+        return ASN1_ERROR_INVALID_STATE;
+
+    ASN1ParseContext & prevContext = mSavedContexts[--mNumSavedContexts];
+
+    if (prevContext.IsIndefiniteLen)
+    {
+        // TODO: implement me
```
[the remark this answers is not in the document]
Comment (emargolis) [reply src/lib/support/ASN1Reader.cpp]: Not required for CHIP

File: src/lib/support/ASN1Writer.cpp [hunk 11]
```diff
@@ -0,0 +1,568 @@
...
+}
+
+ASN1_ERROR ASN1Writer::Finalize()
+{
+    if (mBuf != nullptr)
+    {
+        uint8_t * compactPoint = mBuf;
+        uint8_t * spanStart    = mBuf;
+
+        for (uint8_t ** listEntry = (uint8_t **) mBufEnd; listEntry > mDeferredLengthList;)
```
[the remark this answers is not in the document]
Comment (emargolis) [reply src/lib/support/ASN1Writer.cpp]: done

File: src/lib/support/ASN1.h [hunk 12]
```diff
@@ -0,0 +1,217 @@
...
+
+private:
+    uint8_t * mBuf;
+    uint8_t * mBufEnd;
+    uint8_t * mWritePoint;
+    uint8_t ** mDeferredLengthList;
+
+    ASN1_ERROR EncodeHead(uint8_t cls, uint32_t tag, bool isConstructed, int32_t len);
+    ASN1_ERROR WriteDeferredLength(void);
+    static uint8_t GetLengthOfLength(int32_t len);
```
[the remark this answers is not in the document]
Comment (emargolis) [reply src/lib/support/ASN1.h]: renamed

File: src/lib/support/ASN1.h (hunk 5 above)
Comment (emargolis) [reply src/lib/support/ASN1.h]: done

**Review by mspang — APPROVED** [review]
Reminder to press the "re-request review" button after updating the PR. Otherwise it doesn't show up in the reviewer's "awaiting review" list again.

**Comment by mspang** [comment]
> Should it be in src/crypto ?

^ still wondering about this

**Comment by emargolis** [comment]
> > Should it be in src/crypto ?
> 
> ^ still wondering about this

That was my impression that these ASN1 functions belong to "support" library. These APIs will mostly be used by Crypto, CHIP_Certs, and CHIP_Certs_Tool. I don't know if CHIP_Cert and CHIP_Cert_Tool belong to the "src/crypto". I was thinking to create a separate directory for that "src/credentials" or use "src/lib/protocols/security". This is why I keep ASN1 functions separate in "support"
Your thoughts?

**Comment by gerickson** [comment]
> > > Should it be in src/crypto ?
> > 
> > 
> > ^ still wondering about this
> 
> That was my impression that these ASN1 functions belong to "support" library. These APIs will mostly be used by Crypto, CHIP_Certs, and CHIP_Certs_Tool. I don't know if CHIP_Cert and CHIP_Cert_Tool belong to the "src/crypto". I was thinking to create a separate directory for that "src/credentials" or use "src/lib/protocols/security". This is why I keep ASN1 functions separate in "support"
> Your thoughts?

I think lib/support or lib/asn1 would be appropriate locations since ASN.1 is not crypto-centric.

**Review by rwalker-apple — APPROVED** [review]
a move to another location is fine with me, as well

**Comment by mspang** [comment]
> > > > Should it be in src/crypto ?
> > > 
> > > 
> > > ^ still wondering about this
> > 
> > 
> > That was my impression that these ASN1 functions belong to "support" library. These APIs will mostly be used by Crypto, CHIP_Certs, and CHIP_Certs_Tool. I don't know if CHIP_Cert and CHIP_Cert_Tool belong to the "src/crypto". I was thinking to create a separate directory for that "src/credentials" or use "src/lib/protocols/security". This is why I keep ASN1 functions separate in "support"
> > Your thoughts?
> 
> I think lib/support or lib/asn1 would be appropriate locations since ASN.1 is not crypto-centric.

src/lib/support is the lowest level subcomponent of libCHIP. By putting it here we are implying a dependency from the entire library on ASN.1 support. Not everything in the codebase requires ASN.1 support

Maybe src/crypto is the wrong place, but src/lib/support seems overbroad to me. We should be able to depend on our malloc() replacement without also depending on ASN.1.

**Comment by mspang** [comment]
> > > > > Should it be in src/crypto ?
> > > > 
> > > > 
> > > > ^ still wondering about this
> > > 
> > > 
> > > That was my impression that these ASN1 functions belong to "support" library. These APIs will mostly be used by Crypto, CHIP_Certs, and CHIP_Certs_Tool. I don't know if CHIP_Cert and CHIP_Cert_Tool belong to the "src/crypto". I was thinking to create a separate directory for that "src/credentials" or use "src/lib/protocols/security". This is why I keep ASN1 functions separate in "support"
> > > Your thoughts?
> > 
> > 
> > I think lib/support or lib/asn1 would be appropriate locations since ASN.1 is not crypto-centric.
> 
> src/lib/support is the lowest level subcomponent of libCHIP. By putting it here we are implying a dependency from the entire library on ASN.1 support. Not everything in the codebase requires ASN.1 support
> 
> Maybe src/crypto is the wrong place, but src/lib/support seems overbroad to me. We should be able to depend on our malloc() replacement without also depending on ASN.1.

src/lib/asn1 sounds good though

**Comment by emargolis** [comment]
> > > > > > Should it be in src/crypto ?
> > > > > 
> > > > > 
> > > > > ^ still wondering about this
> > > > 
> > > > 
> > > > That was my impression that these ASN1 functions belong to "support" library. These APIs will mostly be used by Crypto, CHIP_Certs, and CHIP_Certs_Tool. I don't know if CHIP_Cert and CHIP_Cert_Tool belong to the "src/crypto". I was thinking to create a separate directory for that "src/credentials" or use "src/lib/protocols/security". This is why I keep ASN1 functions separate in "support"
> > > > Your thoughts?
> > > 
> > > 
> > > I think lib/support or lib/asn1 would be appropriate locations since ASN.1 is not crypto-centric.
> > 
> > 
> > src/lib/support is the lowest level subcomponent of libCHIP. By putting it here we are implying a dependency from the entire library on ASN.1 support. Not everything in the codebase requires ASN.1 support
> > Maybe src/crypto is the wrong place, but src/lib/support seems overbroad to me. We should be able to depend on our malloc() replacement without also depending on ASN.1.
> 
> src/lib/asn1 sounds good though

moved this new ASN1 library to src/lib/asn1

**Comment by emargolis** [comment]
> Added ASN1 support macros needed for X509 to CHIP certificate conversion.
> 
> #### Problem
> To be able to process and operate with CHIP certificates basic ASN.1 encoding/decoding functions are needed.
> 
> #### Summary of Changes
> ```
>   * Implements ASN1Writer and ASN1Reader parsers
>   * Implements mactors to operate with ASN1Writer/ASN1Reader classes
>   * Implements gen_asn1oid.py python script for automatic generation of OID values
>   * Added ZigBee specific attributes for CHIP Certificates:
>     - ChipNodeId
>     - ChipCAId
>     - ChipSoftwarePublisherId
> ```

@saurabhst @hawk248 @BroderickCarlin @chrisdecenzo this PR needs one more approval.

**Review by saurabhst — APPROVED** [review]

**Pull request merged**